tensorly/decomposition/_parafac2.py

```diff
@@ -348,27 +365,59 @@ def parafac_updates(X, w, f):
         factors = parafac_updates(projected_tensor, weights, factors)
 
         if normalize_factors:
-            new_factors = []
-            for factor in factors:
-                norms = T.norm(factor, axis=0)
-                norms = tl.where(
-                    tl.abs(norms) <= tl.eps(factor.dtype),
-                    tl.ones(tl.shape(norms), **tl.context(factors[0])),
-                    norms,
-                )
+            weights, factors = cp_normalize((weights, factors))
+
+        # Start line search if requested.
+        if line_iter:
+            jump = iteration ** (1.0 / acc_pow)
+
+            ls_weights = weights_last + (weights - weights_last) * jump
+            ls_factors = [
+                factors_last[ii] + (factors[ii] - factors_last[ii]) * jump
+                for ii in range(3)
+            ]
+
+            # Clip if the mode should be non-negative
+            if nn_modes:
+                if 0 in nn_modes:
+                    ls_factors[0] = tl.abs(ls_factors[0])
+                if 2 in nn_modes:
+                    ls_factors[2] = tl.abs(ls_factors[2])
+
+            ls_rec_error = _parafac2_reconstruction_error(
+                tensor_slices, (weights, factors, projections), norm_tensor
+            )
```

Comment:
In Rasmus' recent paper, [[1]], he says that we should also update the projection matrices before computing the reconstruction error for the line search. Have you tested that to see if it makes much of a difference?

Here's the quote from [[1]]:
> In all investigated PARAFAC2 algorithms, matrices A, C and the common matrix U are extrapolated. The orthogonal matrix Pk is not to be extrapolated. Instead, after each extrapolation step, the Pk matrices are calculated according to Algorithm 1 - Step 2 and used for calculating the PARAFAC2 loss function error in order to judge the extrapolation step in the loop.

[1]: https://www.sciencedirect.com/science/article/abs/pii/S0169743921000800

Reply:
Yes, there is a call to compute_projections doing this just before. (I fixed this in a more recent commit, so you may have reviewed code without it.) If you don't do this, the line search is almost always rejected, so it is indeed necessary.